[BUG]: Unclosed regular expression query causes dgraph crash

### What version of Dgraph are you using?

v22.0.2(docker):

```
Dgraph version   : v22.0.2
Dgraph codename  : dgraph
Dgraph SHA-256   : a11258bf3352eff0521bc68983a5aedb9316414947719920d75f12143dd368bd
Commit SHA-1     : 55697a4
Commit timestamp : 2022-12-16 23:03:35 +0000
Branch           : release/v22.0.2
Go version       : go1.18.5
jemalloc enabled : true
```

I have confirmed this on the latest master branch as well:

```
Dgraph version   : v23.0.0-beta1-7-g2fc9a99fa
Dgraph codename  : dgraph
Dgraph SHA-256   : d6ad9375ea69f7188c76039c7004b9592d11cb11fb2095d45675c7319a5b26eb
Commit SHA-1     : 2fc9a99fa
Commit timestamp : 2023-03-05 05:41:58 +0530
Branch           : main
Go version       : go1.19.4
jemalloc enabled : true
```

### Tell us a little more about your go-environment?

go version go1.20 linux/amd64

### Have you tried reproducing the issue with the latest release?

Yes

### What is the hardware spec (RAM, CPU, OS)?

* RAM: 16GB
* CPU: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
* OS: Arch Linux

### What steps will reproduce the bug?

```sh
$ git clone --depth 1 https://github.com/dgraph-io/dgo.git -b v2.1.0                                                                                                                                  
$ grpcurl -plaintext -import-path dgo/protos -proto api.proto \
-d '{"query":"query q($name:string){ q(func: regexp(dgraph.type,$name)) {}}", "vars": {"$name": "/"}}' \
dgraph:9080 api.Dgraph/Query
ERROR:
  Code: Unavailable
  Message: error reading from server: EOF
```

### Expected behavior and actual result.

The expected behavior is to return a parse error. The actual result is crash of dgraph alpha server.

### Additional information

logs during the crash:

```
panic: runtime error: slice bounds out of range [1:0]

goroutine 260 [running]:
github.com/dgraph-io/dgraph/dql.parseRegexArgs({0x2df7658, 0x1})
	/tmp/dgraph/dql/parser.go:1640 +0x165
github.com/dgraph-io/dgraph/dql.regExpVariableFilter(0xc0001a5e00, 0x0)
	/tmp/dgraph/dql/parser.go:457 +0x4d
github.com/dgraph-io/dgraph/dql.substituteVariables(0xc00037fa40, 0x0?)
	/tmp/dgraph/dql/parser.go:398 +0x8ec
github.com/dgraph-io/dgraph/dql.ParseWithNeedVars({{0xc0019b1380?, 0xc0019a6840?}, 0xc001998f00?}, {0x0, 0x0, 0x0})
	/tmp/dgraph/dql/parser.go:632 +0xf0d
github.com/dgraph-io/dgraph/edgraph.parseRequest(0xc0019a6840)
	/tmp/dgraph/edgraph/server.go:1566 +0x40d
github.com/dgraph-io/dgraph/edgraph.(*Server).doQuery(0x22fde28?, {0x22fde28, 0xc001999350}, 0xc0019d9878)
	/tmp/dgraph/edgraph/server.go:1314 +0x992
github.com/dgraph-io/dgraph/edgraph.(*Server).QueryNoGrpc(0x22f6c90?, {0x22fde28?, 0xc001998a20?}, 0xc000e6fb60)
	/tmp/dgraph/edgraph/server.go:1216 +0x2f9
github.com/dgraph-io/dgraph/edgraph.(*Server).Query(0x20?, {0x22fde28, 0xc001998a20}, 0x0?)
	/tmp/dgraph/edgraph/server.go:1182 +0x3b
github.com/dgraph-io/dgo/v210/protos/api._Dgraph_Query_Handler.func1({0x22fde28, 0xc001998a20}, {0x1f5ab60?, 0xc000e6fb60})
	/go/pkg/mod/github.com/dgraph-io/dgo/v210@v210.0.0-20210407152819-261d1c2a6987/protos/api/api.pb.go:1693 +0x78
github.com/dgraph-io/dgraph/ee/audit.AuditRequestGRPC({0x22fde28?, 0xc001998a20?}, {0x1f5ab60?, 0xc000e6fb60?}, 0xc0002dea20?, 0x1dcc180?)
	/tmp/dgraph/ee/audit/interceptor_ee.go:79 +0xe7
github.com/dgraph-io/dgo/v210/protos/api._Dgraph_Query_Handler({0x1f23720?, 0x31fd358}, {0x22fde28, 0xc001998a20}, 0xc0019b4e00, 0x206f250)
	/go/pkg/mod/github.com/dgraph-io/dgo/v210@v210.0.0-20210407152819-261d1c2a6987/protos/api/api.pb.go:1695 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000114f00, {0x2306c40, 0xc001ac1a00}, 0xc0010f7d40, 0xc000159080, 0x2dae2b8, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.52.0/server.go:1336 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc000114f00, {0x2306c40, 0xc001ac1a00}, 0xc0010f7d40, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.52.0/server.go:1704 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/go/pkg/mod/google.golang.org/grpc@v1.52.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.52.0/server.go:963 +0x28a
[Sentry] 2023/03/06 11:37:31 Sending fatal event [2d84c90d17b34d66b33e41d70a109a48] to o318308.ingest.sentry.io project: 5208688
[Sentry] 2023/03/06 11:37:31 Buffer flushed successfully.
```

--- a/dql/parser.go
+++ b/dql/parser.go
@@ -1633,10 +1633,16 @@ type regexArgs struct {
 }
 
 func parseRegexArgs(val string) (regexArgs, error) {
-	end := strings.LastIndex(val, "/")
-	if end < 0 {
-		return regexArgs{}, errors.Errorf("Unexpected error while parsing regex arg: %s", val)
+	// it is possible that val comes in a variable and has not been validated yet.
+	// we do the validation here and ensure that it looks like: /<expr>/.
+	if len(val) == 0 {
+		return regexArgs{}, errors.Errorf("Found empty regex: [%s]", val)
 	}
+	end := strings.LastIndex(val, "/")
+	if end < 1 || val[0] != '/' {
+		return regexArgs{}, errors.Errorf("Found invalid regex: [%s]", val)
+	}
+
 	expr := strings.Replace(val[1:end], "\\/", "/", -1)
 	flags := ""
 	if end+1 < len(val) {

--- a/query/query0_test.go
+++ b/query/query0_test.go
@@ -21,7 +21,9 @@ package query
 
 import (
 	"context"
+	"fmt"
 	"os"
+	"strings"
 	"testing"
 
 	"github.com/stretchr/testify/require"
@@ -3510,6 +3512,26 @@ func TestMatchingWithPagination(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			result := processQueryNoErr(t, tc.query)
 			require.JSONEq(t, tc.expected, result)
+		})
+	}
+}
+
+func TestInvalidRegex(t *testing.T) {
+	testCases := []struct {
+		regex  string
+		errStr string
+	}{
+		{"/", "invalid"},
+		{"", "empty"},
+		{"/?", "invalid"},
+		{"=/?", "invalid"},
+		{"aman/", "invalid"},
+	}
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("test%d regex=%v", i, tc.regex), func(t *testing.T) {
+			vars := map[string]string{"$name": tc.regex}
+			_, err := processQueryWithVars(t, `query q($name:string){ q(func: regexp(dgraph.type, $name)) {}}`, vars)
+			require.Contains(t, strings.ToLower(err.Error()), tc.errStr)
 		})
 	}
 }

--- a/systest/mutations_test.go
+++ b/systest/mutations_test.go
@@ -78,6 +78,7 @@ func TestSystem(t *testing.T) {
 	t.Run("has with dash", wrap(HasWithDash))
 	t.Run("list geo filter", wrap(ListGeoFilterTest))
 	t.Run("list regex filter", wrap(ListRegexFilterTest))
+	t.Run("regex query with vars with slash", wrap(RegexQueryWithVarsWithSlash))
 	t.Run("regex query vars", wrap(RegexQueryWithVars))
 	t.Run("graphql var child", wrap(GraphQLVarChild))
 	t.Run("math ge", wrap(MathGe))
@@ -1220,6 +1221,60 @@ func ListRegexFilterTest(t *testing.T, c *dgo.Dgraph) {
 			},
 			{
 				"name": "admin"
+			}
+		]
+	}
+	`, string(resp.GetJson()))
+}
+
+func RegexQueryWithVarsWithSlash(t *testing.T, c *dgo.Dgraph) {
+	ctx := context.Background()
+
+	op := &api.Operation{Schema: `data: [string] @index(trigram) .`}
+	require.NoError(t, c.Alter(ctx, op))
+
+	txn := c.NewTxn()
+	defer func() { require.NoError(t, txn.Discard(ctx)) }()
+	_, err := txn.Mutate(ctx, &api.Mutation{
+		CommitNow: true,
+		SetNquads: []byte(`
+			_:a <data> "abc/def"  .
+			_:b <data> "fgh" .
+			_:c <data> "ijk" .
+		`),
+	})
+	require.NoError(t, err)
+
+	// query without variable
+	resp, err := c.NewTxn().Query(context.Background(), `{
+		q(func: regexp(data, /\/def/)) {
+			data
+		}
+	}`)
+	require.NoError(t, err)
+	testutil.CompareJSON(t, `
+	{
+		"q": [
+			{
+				"data": ["abc/def"]
+			}
+		]
+	}
+	`, string(resp.GetJson()))
+
+	// query with variable
+	resp, err = c.NewTxn().QueryWithVars(context.Background(), `
+	query search($rex: string){
+		q(func: regexp(data, $rex)) {
+			data
+		}
+	}`, map[string]string{"$rex": "/\\/def/"})
+	require.NoError(t, err)
+	testutil.CompareJSON(t, `
+	{
+		"q": [
+			{
+				"data": ["abc/def"]
 			}
 		]
 	}

